ResearchKit/Common/ORKAnswerFormat.m

```diff
@@ -1887,39 +1908,40 @@ - (instancetype)copyWithZone:(NSZone *)zone {
 - (BOOL)isAnswerValid:(id)answer {
     BOOL isValid = NO;
     if ([answer isKindOfClass:[NSString class]]) {
-        return [self isAnswerValidWithString:(NSString *)answer];
+        isValid = [self isAnswerValidWithString:(NSString *)answer];
     }
     return isValid;
 }
 
 - (BOOL)isAnswerValidWithString:(NSString *)text {
-    if (text.length > 0) {
-        return ([self isTextLengthValidWithString:text] && [self isEmailAddressValidWithString:text]);
+    if (text) {
+        BOOL isValid = NO;
+        if (self.regex) {
+            NSString *regExPattern = self.regex;
+            NSRegularExpression *regEx = [[NSRegularExpression alloc] initWithPattern:regExPattern options:NSRegularExpressionCaseInsensitive error:nil];
+            NSUInteger regExMatches = [regEx numberOfMatchesInString:text options:0 range:NSMakeRange(0, [text length])];
+            isValid = (regExMatches != 0);
+        } else if (text.length > 0) {
+	        isValid = [self isTextLengthValidWithString:text];
+    	} else {
+            isValid = YES;
+        }
+        return isValid;
+    } else {
+        return YES;
     }
-    
-    return YES;
 }
 
 - (BOOL)isTextLengthValidWithString:(NSString *)text {
     return (_maximumLength == 0 || text.length <= _maximumLength);
 }
 
-- (BOOL)isEmailAddressValidWithString:(NSString *)text {
-    if (self.isEmailAddress) {
-        NSPredicate *emailValidationTest = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", EmailValidationRegex];
-        return [emailValidationTest evaluateWithObject:text];
-    }
-    
-    return YES;
-}
-
 - (NSString *)localizedInvalidValueStringWithAnswerString:(NSString *)text {
     NSString *string = nil;
     if (![self isTextLengthValidWithString:text]) {
         string = [NSString stringWithFormat:ORKLocalizedString(@"TEXT_ANSWER_EXCEEDING_MAX_LENGTH_ALERT_MESSAGE", nil), ORKLocalizedStringFromNumber(@(_maximumLength))];
```

Comment:
We need to include a case to show the `invalidMessage` property. Firstly, we want to check if the textLength is valid. If not, then we want to append it to `string`. Then next we want to check if the text complies to the `regex`. If not, then we want to append that message to `string` as well and return it.

To achieve this, we need to do something along the lines of the following:
```
- (NSString *)localizedInvalidValueStringWithAnswerString:(NSString *)text {
    NSMutableString *string = [NSMutableString new];
    if (![self isTextLengthValidWithString:text]) {
        [string appendString:[NSString stringWithFormat:ORKLocalizedString(@"TEXT_ANSWER_EXCEEDING_MAX_LENGTH_ALERT_MESSAGE", nil), ORKLocalizedStringFromNumber(@(_maximumLength))]];
    }
    if (![self isTextRegexValidWithString:text]) {
        [string appendString:_invalidMessage];
    }
    return string;
}
```

***Currently, if an invalid input is entered that doesn't match the regex, no alert is shown, because this method is returning nil.***